sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -84,19 +84,49 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
     }
   }
 
+  // Extract a list of logical plans to be joined for join-order comparisons.
+  // Since `ExtractFiltersAndInnerJoins` handles left-deep trees only, this function have
+  // the same strategy to extract the plan list.
+  private def extractLeftDeepInnerJoins(plan: LogicalPlan): Seq[LogicalPlan] = plan match {
+    case j @ Join(left, right, _: InnerLike, _) => right +: extractLeftDeepInnerJoins(left)
+    case p @ Project(_, j @ Join(_, _, _: InnerLike, _)) => extractLeftDeepInnerJoins(j)
+    case _ => Seq(plan)
+  }
+
+  private def checkSameJoinOrder(plan1: LogicalPlan, plan2: LogicalPlan): Boolean = {
+    extractLeftDeepInnerJoins(plan1) == extractLeftDeepInnerJoins(plan2)
+  }
+
+  private def mayCreateOrderedJoin(
+      originalPlan: LogicalPlan,
+      input: Seq[(LogicalPlan, InnerLike)],
+      conditions: Seq[Expression]): LogicalPlan = {
+    val orderedJoins = createOrderedJoin(input, conditions)
+    if (!checkSameJoinOrder(orderedJoins, originalPlan)) {
```

Comment:
Is this check necessary? I think check `originalPlan.output != orderedJoins.output` is enough, and faster.

Reply:
If we don't have this check, `operatorOptimizationRuleSet` reaches `fixedPoint` because `ReorderJoin` is re-applied in the same join trees every time the optimization rule batch invoked. This case does not happen in the master because reordered joins have `Project` in internal nodes (`Project` added by following optimization rules, e.g., `ColumnPruning`) and this plan structure guards this case.

Reply:
ah, right, thanks!